pkg/appcommon/harness.go

```diff
@@ -112,13 +113,16 @@ func New(cfg Config, reg prometheus.Registerer, metricPrefix string) (app App, e
 		return app, fmt.Errorf("can't initialize the instrumentation middleware %w", err)
 	}
 
-	tracer, closer, err := NewTracer(cfg.ServiceName, logger)
-	if err != nil {
-		return app, err
+	app.Tracer = cfg.Tracer
+	if cfg.Tracer == nil {
+		tracer, closer, err := NewTracer(cfg.ServiceName, logger)
+		if err != nil {
+			return app, err
+		}
+		app.closers = append(app.closers, closer.Close)
+		app.Tracer = tracer
```

Comment:
Slightly more idiomatic way of doing this -- this way we never assign the invalid value to app.Tracer. In general I try to avoid the pattern of "set value and then check to see if the value is actually ok."

```suggestion
	if cfg.Tracer != nil {
		app.Tracer = cfg.Tracer
	} else {
		tracer, closer, err := NewTracer(cfg.ServiceName, logger)
		if err != nil {
			return app, err
		}
		app.closers = append(app.closers, closer.Close)
		app.Tracer = tracer
```